Homework/lesson 18/task1.py

```diff
@@ -0,0 +1,139 @@
+# task1
+import re
+from functools import wraps
+
+
+class MyClass:
+
+    def __init__(self, email):
+        self.email = email
+        self.validate(email)
+
+    @classmethod
+    def validate(cls, email):
+        if not re.match(r"[^@]+@[^@]+\.[^@]+", email):
+            raise ValueError("Invalid email address")
+
+
+mail = MyClass("123@345.com")
+print(mail.email)
+
+
+# mail = MyClass("123a345.email")
+
+# task2
+
+
+class Boss:
+
+    def __init__(self, id_: int, name: str, company: str):
+        self.id = id_
+        self.name = name
+        self.company = company
+        self._workers = []
+
+    @property
+    def workers(self):
+        return self._workers
+
+    def add_worker(self, worker):
+        if not isinstance(worker, Worker):
+            raise TypeError("Can only add instances of Worker class to the workers list")
+        self._workers.append(worker)
+
+
+class Worker:
+
+    def __init__(self, id_: int, name: str, company: str, boss: Boss):
+        self.id = id_
+        self.name = name
+        self.company = company
+        self._boss = None
+        self.boss = boss
+
+    @property
+    def boss(self):
+        return self._boss
+
+    @boss.setter
+    def boss(self, new_boss):
+        if not isinstance(new_boss, Boss):
+            raise TypeError("A worker's boss must be an instance of the Boss class")
+        if self._boss is not None:
```

Comment:
It's not gonna be working because of above's condition: `if not isinstance(new_boss, Boss):` , is it?

Reply:
But this is really nice usecase!

Reply:
Sorry, I didn't get question